cluster-autoscaler/core/scaledown/actuation/delete_in_batch.go

```diff
@@ -0,0 +1,173 @@
+	MaxKubernetesEmptyNodeDeletionTime = 3 * time.Minute
+	// MaxCloudProviderNodeDeletionTime is the maximum time needed by cloud provider to delete a node.
+	MaxCloudProviderNodeDeletionTime = 5 * time.Minute
+)
+
+// NodeBatcherDeleterImpl implements NodeBatcherDeleter interface.
+type NodeBatcherDeleterImpl struct {
+	sync.Mutex
+	ctx                   *context.AutoscalingContext
+	clusterState          *clusterstate.ClusterStateRegistry
+	nodeDeletionTracker   *deletiontracker.NodeDeletionTracker
+	deletionsPerNodeGroup map[string][]*apiv1.Node
+}
+
+// NewNodeBatcherDeleter return new NodeBatchDeleter
+func NewNodeBatcherDeleter(ctx *context.AutoscalingContext, csr *clusterstate.ClusterStateRegistry, nodeDeletionTracker *deletiontracker.NodeDeletionTracker) deletionbatcher.NodeBatcherDeleter {
+	return &NodeBatcherDeleterImpl{
+		ctx:                   ctx,
+		clusterState:          csr,
+		nodeDeletionTracker:   nodeDeletionTracker,
+		deletionsPerNodeGroup: make(map[string][]*apiv1.Node),
+	}
+}
+
+// AddToBucket adds node to delete candidates.
+func (d *NodeBatcherDeleterImpl) AddToBucket(node *apiv1.Node) (string, error) {
+	d.Lock()
+	defer d.Unlock()
+	nodeGroup, err := d.ctx.CloudProvider.NodeGroupForNode(node)
+	if err != nil {
+		return "", err
+	}
+	val, ok := d.deletionsPerNodeGroup[nodeGroup.Id()]
+	if !ok || len(val) == 0 {
+		d.deletionsPerNodeGroup[nodeGroup.Id()] = []*apiv1.Node{node}
+		return nodeGroup.Id(), nil
+	}
+	d.deletionsPerNodeGroup[nodeGroup.Id()] = append(d.deletionsPerNodeGroup[nodeGroup.Id()], node)
+	return "", nil
+}
+
+// CleanUp delete nodes of a given nodeGroup.
+func (d *NodeBatcherDeleterImpl) CleanUp(nodeGroupId string) error {
+	d.Lock()
+	defer d.Unlock()
+	nodes, ok := d.deletionsPerNodeGroup[nodeGroupId]
+	if !ok {
+		return nil
+	}
+	delete(d.deletionsPerNodeGroup, nodeGroupId)
+
+	go func(nodes []*apiv1.Node, nodeGroupId string) {
+		var result status.NodeDeleteResult
+		err := d.deleteNodesFromCloudProvider(nodes)
+		if err != nil {
+			result = status.NodeDeleteResult{ResultType: status.NodeDeleteErrorFailedToDelete, Err: err}
+		} else {
+			result = status.NodeDeleteResult{ResultType: status.NodeDeleteOk}
+		}
+		for _, node := range nodes {
+			if d.nodeDeletionTracker.IsNodeEmpty(node.Name) {
+				d.processEmptyNode(node, nodeGroupId, result)
+			} else if d.nodeDeletionTracker.IsNodeDrained(node.Name) {
+				d.processNonEmptyNode(node, nodeGroupId, result)
+			} else {
+				klog.Errorf("Node %s is not present in nodeDeletionTracker", node.Name)
+			}
+		}
+	}(nodes, nodeGroupId)
+	return nil
+}
+
+// deleteNodeFromCloudProvider removes the given nodes from cloud provider. No extra pre-deletion actions are executed on
+// the Kubernetes side. If successful, the deletion is recorded in CSR, and an event is emitted on the node.
+func (d *NodeBatcherDeleterImpl) deleteNodesFromCloudProvider(nodes []*apiv1.Node) error {
+	d.Lock()
```

Comment:
This method doesn't modify `d.deletionsPerNodeGroup`, so it shouldn't need a lock, right?